### ISSUE TITLE
RNN that chooses to act, on episodes with multiple trials

### ISSUE THREAD

**Issue by mobeets** [mobeets/sarsa-rnn#10]
Okay, after #9, I now have an RNN that integrates the stimulus, and chooses to act after an integration time that is longer for noisier stimuli. Each episode was a single trial, and I set `{'fail': -5}`.

The goal now is to train on episodes consisting of _multiple_ trials, with no penalty for failing.

Steps:
- Add in an ITI between trials (geometric distribution with a minimum time)
- Concatenate trials with the penalty for failing, and confirm we still get RTs that scale with 1/coh
- Try removing the penalty for failing; potentially add a temporal penalty for failing

**Comment by mobeets** [comment]
Okay I have done the first two steps above. The thing I'm struggling with is that `{'fail': -5}` seems necessary still. Anyway, what I have here is a network trained on single-trial episodes that could then very easily be deployed on multi-trial episodes (5 trials per episode). $\gamma = 0.98$, $h=5$, batch size 1. Trained only one one coherence level (12.8) with `sigma=2`.

Here's its performance, averaged across the trial index within an episode:

<img width="600px" src="https://user-images.githubusercontent.com/1677179/182455717-af43fe69-e1d3-4aa7-953e-a4f5a065bedb.png"/>

<img width="600px" src="https://user-images.githubusercontent.com/1677179/182455875-313681c7-a2ac-4765-ab0e-c8e69aa524e4.png"/>

<img width="300px" src="https://user-images.githubusercontent.com/1677179/182455918-90bc9cd2-d1ad-471c-95e4-402a827e0ecf.png"/>

So basically what it's doing is that its policy never gets to 1 for the correct action, but it does make sure that the incorrect action has 0 probability, so with enough time steps it will eventually give the correct answer. I suspect that the penalty for failing is partly why this is happening.

**Comment by mobeets** [comment]
Confirming my hunch, if I lower `{'fail': -1}` and continue training the above network, this is what happens:

<img width="600px" src="https://user-images.githubusercontent.com/1677179/182456934-68591173-87d9-4ddb-ac3b-68e332cd788e.png"/>
<img width="600px" src="https://user-images.githubusercontent.com/1677179/182456993-2c36b879-2a66-415d-b5c0-e959af58c5c0.png"/>

So it acts sooner, since there's now less of a penalty for it.

**Comment by mobeets** [comment]
To do tomorrow: __add a temporal penalty (i.e., a "timeout") to incorrect trials__ (somehow).

This is likely going to have to be implemented as some sort of wrapper. Possibly we want to increment the next trial's `iti` based on an incorrect response. Seems tricky.

See [here](https://github.com/neurogym/neurogym/blob/master/neurogym/wrappers/side_bias.py).

So yeah if I can a) augment the task so that the trial parameters (in _new_trial()) to accept an iti, I can then write a wrapper around new_trial that provides an initial duration as a function of the last trial's correctness. The trick is, it's not clear yet to me how to dynamically change the length of a period. It is sampled at some point, so I guess I'll have to figure out how that works. 

It turns out I'll have to change `self.timing`, since in new_trial we call `self.add_period` which then uses the timings.

**Comment by mobeets** [comment]
I just realized, for a timeout to work, we need an observation not of the reward (as in the current wrapper), but an observation of correct vs incorrect. Ie so that even if there is no penalty for failure, there is an observation of it. This is the only way the network could learn to avoid missing trials prematurely 

So in the future let's only use a correctness wrapper (which I will have to make myself somehow)
